vm/compiler/src/coinbase_puzzle/helpers/epoch_challenge/mod.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+mod bytes;
+
+use snarkvm_algorithms::fft::Evaluations as EvaluationsOnDomain;
+
+use super::*;
+use crate::coinbase_puzzle::hash_to_polynomial;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct EpochChallenge<N: Network> {
+    /// The epoch number.
+    epoch_number: u32,
+    /// The epoch block hash, defined as the block hash right before the epoch updated.
+    epoch_block_hash: N::BlockHash,
+    /// The epoch polynomial.
+    epoch_polynomial: DensePolynomial<<N::PairingCurve as PairingEngine>::Fr>,
+    /// The evaluations of the epoch polynomial over the product domain.
+    epoch_polynomial_evaluations: EvaluationsOnDomain<<N::PairingCurve as PairingEngine>::Fr>,
+}
+
+impl<N: Network> EpochChallenge<N> {
+    /// Initializes a new epoch challenge.
+    pub fn new(epoch_number: u32, epoch_block_hash: N::BlockHash, degree: u32) -> Result<Self> {
+        // Construct the 'input' as '( epoch_number || epoch_block_hash )'
+        let input: Vec<u8> =
+            epoch_number.to_le_bytes().into_iter().chain(epoch_block_hash.to_bytes_le()?.into_iter()).collect();
+        let epoch_polynomial = hash_to_polynomial::<<N::PairingCurve as PairingEngine>::Fr>(&input, degree)?;
+        let domain = EvaluationDomain::new((2 * degree - 1) as usize).unwrap();
```

Comment:
note: `degree` can't be equal to `0`; if this is guaranteed, feel free to mark this comment as resolved

Reply:
Good point; we should take a `NonZeroU32` as input.

Reply:
I've handled this by `ensure`-ing that `degree` non-zero in [`ef950de` (#1100)](https://github.com/AleoHQ/snarkVM/pull/1100/commits/ef950dea9ab89461d8f6e1768b5d303ee7cd66f7)